extensions/amp-font/0.1/amp-font.js

```diff
@@ -0,0 +1,327 @@
+
+
+  /**
+   * Compare dimensions between elements styled with default fonts and custom
+   * font.
+   * @returns {boolean} Returns true if the dimensions are noticeably different
+   * else returns false.
+   * @private
+   */
+  compareMeasurements_() {
+    let hasWidthChanged = false;
+    let hasHeightChanged = false;
+    this.domElements_.DEFAULT_FONT_ELEMENTS.some(defaultElement => {
+      hasWidthChanged = (
+          Math.abs(
+              defaultElement./*REVIEW*/offsetWidth -
+              this.domElements_.CUSTOM_FONT_ELEMENT./*REVIEW*/offsetWidth) >
+            TOLERANCE_);
+      hasWidthChanged = (
+          Math.abs(
+              defaultElement./*REVIEW*/offsetHeight -
+              this.domElements_.CUSTOM_FONT_ELEMENT./*REVIEW*/offsetHeight) >
+            TOLERANCE_);
+      if (hasWidthChanged || hasHeightChanged) {
+        return true;
+      }
+    });
+    if (hasWidthChanged || hasHeightChanged) {
+      return true;
+    }
+    return false;
+  }
+
+
+  /**
+   * @private
+   */
+  onFontLoadSuccess_() {
+    timer.cancel(this.timeoutHandle_);
+    var addClassName = this.element.getAttribute('on-load-add-class');
+    var removeClassName =
+        this.element.getAttribute('on-load-remove-class');
+    if (addClassName) {
+      this.documentElement_.classList.add(addClassName);
+    }
+    if (removeClassName) {
+      this.documentElement_.classList.remove(removeClassName);
+      this.document_.body.classList.remove(removeClassName);
+    }
+    this.dispose_();
+  }
+
+
+  /**
+   * @private
+   */
+  onFontLoadError_() {
+    var addClassName = this.element.getAttribute('on-error-add-class');
+    var removeClassName =
+        this.element.getAttribute('on-error-remove-class');
+    if (addClassName) {
+      this.documentElement_.classList.add(addClassName);
+    }
+    if (removeClassName) {
+      this.documentElement_.classList.remove(removeClassName);
+      this.document_.body.classList.remove(removeClassName);
+    };
+    this.dispose_();
+  }
+
+
+  /**
+   * @private
+   */
+  dispose_() {
+    if (this.domElements_.CONTAINER) {
+      removeElement(this.domElements_.CONTAINER);
+    }
+    this.domElements_ = {
+      CONTAINER: null,
+      DEFAULT_FONT_ELEMENTS: [],
+      CUSTOM_FONT_ELEMENT: null
+    };
+    if (this.timeoutHandle_) {
+      timer.cancel(this.timeoutHandle_);
+    }
+  }
+
+
+  /**
+   * Starts the timeout counter to not attempt loading the font after the
+   * timeout normal.
+   * @private
+   */
+  beginTimeout_() {
+    let timeoutAttribute = this.element.getAttribute('timeout');
+    let timeoutInMs = parseInt(timeoutAttribute, 10);
+    timeoutInMs = isNaN(timeoutInMs) || timeoutInMs < 0 ?
+        DEFAULT_TIMEOUT_ : timeoutInMs;
+    let timeSinceStart = timer.now() - this.startTime_;
+    timeoutInMs = timeSinceStart > timeoutInMs ?
+        0 : (timeoutInMs - timeSinceStart);
+    this.timeoutHandle_ = timer.delay(() => {
+      this.hasTimedOut_ = true;
```

Comment:
Are we always guaranteed to be able to cancel timeout in time? E.g. is a risk condition possible where we got a success signal and then timer is still called? Maybe we also need `hasSucceeded_` state var?